src/hooks/FreelyTransferable.sol

```diff
@@ -0,0 +1,171 @@
+        }
+
+        return true;
+    }
+
+    //----------------------------------------------------------------------------------------------
+    // Restriction updates
+    //----------------------------------------------------------------------------------------------
+
+    /// @inheritdoc ITransferHook
+    function updateRestriction(address token, bytes memory payload) external auth {
+        UpdateRestrictionType updateId = payload.updateRestrictionType();
+
+        if (updateId == UpdateRestrictionType.Member) {
+            UpdateRestrictionMessageLib.UpdateRestrictionMember memory m = payload.deserializeUpdateRestrictionMember();
+            updateMember(token, m.user.toAddress(), m.validUntil);
+        } else if (updateId == UpdateRestrictionType.Freeze) {
+            UpdateRestrictionMessageLib.UpdateRestrictionFreeze memory m = payload.deserializeUpdateRestrictionFreeze();
+            freeze(token, m.user.toAddress());
+        } else if (updateId == UpdateRestrictionType.Unfreeze) {
+            UpdateRestrictionMessageLib.UpdateRestrictionUnfreeze memory m =
+                payload.deserializeUpdateRestrictionUnfreeze();
+            unfreeze(token, m.user.toAddress());
+        } else {
+            revert InvalidUpdate();
+        }
+    }
+
+    /// @inheritdoc IFreezable
+    function freeze(address token, address user) public auth {
+        require(user != address(0), CannotFreezeZeroAddress());
+        require(!root.endorsed(user), EndorsedUserCannotBeFrozen());
+
+        uint128 hookData = uint128(IShareToken(token).hookDataOf(user));
+        IShareToken(token).setHookData(user, bytes16(hookData.setBit(FREEZE_BIT, true)));
+
+        emit Freeze(token, user);
+    }
+
+    /// @inheritdoc IFreezable
+    function unfreeze(address token, address user) public auth {
+        uint128 hookData = uint128(IShareToken(token).hookDataOf(user));
+        IShareToken(token).setHookData(user, bytes16(hookData.setBit(FREEZE_BIT, false)));
+
+        emit Unfreeze(token, user);
+    }
+
+    /// @inheritdoc IFreezable
+    function isFrozen(address token, address user) public view returns (bool) {
+        return uint128(IShareToken(token).hookDataOf(user)).getBit(FREEZE_BIT);
+    }
+
+    /// @inheritdoc IMemberlist
+    function updateMember(address token, address user, uint64 validUntil) public auth {
+        require(block.timestamp <= validUntil, InvalidValidUntil());
+        require(!root.endorsed(user), EndorsedUserCannotBeUpdated());
+
+        uint128 hookData = uint128(validUntil) << 64;
+        hookData.setBit(FREEZE_BIT, isFrozen(token, user));
+        IShareToken(token).setHookData(user, bytes16(hookData));
+
+        emit UpdateMember(token, user, validUntil);
+    }
+
+    /// @inheritdoc IMemberlist
+    function isMember(address token, address user) external view returns (bool isValid, uint64 validUntil) {
+        validUntil = abi.encodePacked(IShareToken(token).hookDataOf(user)).toUint64(0);
```

Comment:
Not a blocker:

`validUntil` is obtained using both `>> 64` and with `toUint64(0)`. Maybe we can normalize or isolate that into a method/lib?

Reply:
Added as generic item to the v3.1 board ([ref](https://www.notion.so/Clean-up-hooks-2302eac24e178008b016dddb9ee95f84?source=copy_link))

Reply:
I had already added `Implement BaseTransferHook` btw, just not in the v3.1 list